src/test/PulsarMock.js

```diff
@@ -0,0 +1,229 @@
+              resolve(messages[messageIndex].message);
+            } else {
+              // No visible messages available, wait for new messages
+              if (!PulsarClientMock.listeners.has(topic)) {
+                PulsarClientMock.listeners.set(topic, []);
+              }
+              PulsarClientMock.listeners.get(topic).push(tryResolve);
+            }
+          };
+
+          tryResolve();
+        });
+      },
+      acknowledge: async (message) => {
+        const resolveLock = await PulsarClientMock.lockOperation(topic); // Lock operation for topic
+        try {
+          await PulsarClientMock.acknowledgeMessage(topic, message.messageId.id);
+        } finally {
+          resolveLock(); // Unlock operation for topic
+          PulsarClientMock.operationLocks.delete(topic); // Cleanup lock
+        }
+      },
+      acknowledgeId: async (messageId) => {
+        const resolveLock = await PulsarClientMock.lockOperation(topic); // Lock operation for topic
+        try {
+          await PulsarClientMock.acknowledgeMessage(topic, messageId.id);
+        } finally {
+          resolveLock(); // Unlock operation for topic
+          PulsarClientMock.operationLocks.delete(topic); // Cleanup lock
+        }
+      },
+      close: async () => {},
+    };
+  }
+
+  static async acknowledgeMessage(topic, messageId) {
+    // No need to lock here since it's already locked in acknowledge and acknowledgeId methods
+    const messages = PulsarClientMock.inMemoryStore.get(topic) || [];
+    const messageIndex = messages.findIndex(m => m.message.messageId.id === messageId);
+    if (messageIndex !== -1) {
+      const [acknowledgedMessage] = messages.splice(messageIndex, 1);
+      if (!this.acknowledgedMessages.has(topic)) {
+        this.acknowledgedMessages.set(topic, []);
+      }
+      this.acknowledgedMessages.get(topic).push(acknowledgedMessage.message);
+    }
+  }
+
+
+  static getTopics() {
+    return Array.from(PulsarClientMock.inMemoryStore.keys());
+  }
+
+  // Method to return statistics for the client
+  static getStats(topic) {
+    const messages = this.inMemoryStore.get(topic) || [];
+    const acknowledged = this.acknowledgedMessages.get(topic) || [];
+    const inFlight = messages.filter(m => !m.visible).length;
+    const queueLength = messages.length + acknowledged.length;
+
+    return {
+      acknowledgedCount: acknowledged.length,
+      inFlightCount: inFlight,
+      queueLength,
+    };
+  }
+
+  // Method to return all acknowledged messages for a topic
+  static getAcknowledgedMessages(topic) {
+    return this.acknowledgedMessages.get(topic) || [];
+  }
+
+  static clear() {
+    PulsarClientMock.inMemoryStore.clear();
+    PulsarClientMock.listeners.clear();
+    PulsarClientMock.acknowledgedMessages.clear();
+  }
+
+  static notifyListeners(topic) {
+    const listeners = PulsarClientMock.listeners.get(topic) || [];
+    while (listeners.length > 0) {
+      const listener = listeners.shift();
```

Comment:
why should notifying listeners involve removing the listeners.

Reply:
Added comments. Listeners are actually receive invocations, which were blocked because no messages were available at the moment.